File: C-Java/tp2/partieB.c
////////////////////////////////////////////
//  Travail pratique #2                   //
//  Auteur : Patrick Chartrand (20017273) //
//  Date : 9 mars 2020                    //
////////////////////////////////////////////


// PARTIE B
#include <stdio.h>

int main()
{

  char poste[8] = { 'P', 'P', 'O', 'A', 'P', 'A', 'P', 'P' };
  int cafe[8] = {  3 ,  1 ,  6 ,  0 ,  5 ,  4 ,  0 ,  3  };
  int age[8] = { 25 , 19 , 27 , 26 , 69 , 24 , 56 , 29  };
 
  int nbPers = sizeof(age) / sizeof(int),
    // variable pour les boucles for et le tri
    i,
    j, 
    indMin,
    // variables temporaires
    tempoAge,
    tempoCafe, 
    tempoPoste;


  // AFFICHAGE DU TABLEAU 1
  // titre
  printf("Tableau 1 :\npostes, consommations de café et âges\n");
  printf("===========================================================\n");
  // données par poste
  printf("| Postes |");     
  for (i = 0; i < nbPers ; i++)
  printf("|  %c  ", poste[i]);
  printf("|\n");
  printf("-----------------------------------------------------------\n");      

  // données par quantité de café
  printf("| Cafés  |");     
  for (i = 0; i < nbPers ; i++)
  printf("|  %d  ", cafe[i]);
  printf("|\n");
  printf("-----------------------------------------------------------\n");      

  // données par âges
  printf("|  Âges  |");     
  for (i = 0; i < nbPers ; i++)
  printf("|  %d ", age[i]);
  printf("|\n");
  printf("===========================================================\n\n\n");      


  // CALCULS À PARTIR DES DONNÉES DU TABLEAU 1
  // calcul du nombre total de programmeurs
  j = 0;
  for (i = 0; i < nbPers; i++)
  {
    if (poste[i] == 'P') j++;
  }
  printf("Nombre de programmeurs : %d\n", j);

  // calcul du nombre total de secrétaires
  j = 0;
  for (i = 0; i < nbPers; i++) 
  {
    if (poste[i] == 'S') j++;
  }
  printf("Nombre de secrétaires : %d\n", j);

  // calcul de la plus petite consommation de café
  j = 100;
  for (i = 0; i < nbPers; i++)
  {
    if (poste[i] == 'A' && cafe[i] < j)
    j = cafe[i];
  }
  printf("Nombre minimale de consommation en café par analyste : %d\n", j);

  // calcul du plus grand âge parmi les analystes
  j = 0;
  for (i = 0; i < nbPers; i++)
  {
    if (poste[i] == 'P' && age[i] > j)
    j = age[i];
  }
  printf("Plus grand âge parmi tous les programmeurs : %d\n", j);

  // calcul de la consommation moyenne de café par les programmeurs
  j = 0;
  for (i = 0; i < nbPers; i++)
  if (poste[i] == 'O')
  {
    j += cafe[i];
  }
  printf("Moyenne de la consommation en café chez les opérateurs : %0.2f\n", ((float)j / nbPers));

  // calcul de l'âge moyen de tous les employés
  j = 0;
  for (i = 0; i < nbPers; i++)
  {
    j += age[i];
  }
  printf("Moyenne d'âge parmi tous les employés : %0.2f ans\n\n\n", ((float)j / nbPers));
 

  // TRI DES DONNÉES DU TABLEAU 1
  // Le tri par âge
  for (i = 0; i < nbPers - 1; i++)
  {
    indMin = i;
    for (j = i + 1; j < nbPers; j++)
    if (age[j] < age[indMin])
    indMin = j;
    if (indMin != i)
    {
      tempoAge = age[i];
      age[i] = age[indMin];
      age[indMin] = tempoAge;
          
      tempoCafe = cafe[i];
      cafe[i] = cafe[indMin];
      cafe[indMin] = tempoCafe;

      tempoPoste = poste[i];
      poste[i] = poste[indMin];
      poste[indMin] = tempoPoste;
    }      
  }


  // AFFICHAGE DU TABLEAU 2
  // titre
  printf("Tableau 2 :\npostes, consommations de café par ordre croissant d'âges\n");   
  printf("===========================================================\n");
  // données par poste    
  printf("| Postes |");
  for (i = 0; i < nbPers ; i++)
  printf("|  %c  ", poste[i]);
  printf("|\n");
  printf("-----------------------------------------------------------\n");      

  // données par quantité de café
  printf("| Cafés  |");     
  for (i = 0; i < nbPers ; i++)
  printf("|  %d  ", cafe[i]);
  printf("|\n");
  printf("-----------------------------------------------------------\n");      

  // données par âges
  printf("|  Âges  |");     
  for (i = 0; i < nbPers ; i++)
  printf("|  %d ", age[i]);
  printf("|\n");
  printf("===========================================================\n\n\n");


  return 0;

}

/* EXÉCUTION

Tableau 1 :
postes, consommations de café et âges
===========================================================
| Postes ||  P  |  P  |  O  |  A  |  P  |  A  |  P  |  P  |
-----------------------------------------------------------
| Cafés  ||  3  |  1  |  6  |  0  |  5  |  4  |  0  |  3  |
-----------------------------------------------------------
|  Âges  ||  25 |  19 |  27 |  26 |  69 |  24 |  56 |  29 |
===========================================================


Nombre de programmeurs : 5
Nombre de secrétaires : 0
Nombre minimale de consommation en café par analyste : 0
Plus grand âge parmi tous les programmeurs : 69
Moyenne de la consommation en café chez les opérateurs : 0.75
Moyenne d'âge parmi tous les employés : 34.38 ans


Tableau 2 :
postes, consommations de café par ordre croissant d'âges
===========================================================
| Postes ||  P  |  A  |  P  |  A  |  O  |  P  |  P  |  P  |
-----------------------------------------------------------
| Cafés  ||  1  |  4  |  3  |  0  |  6  |  3  |  0  |  5  |
-----------------------------------------------------------
|  Âges  ||  19 |  24 |  25 |  26 |  27 |  29 |  56 |  69 |
===========================================================

*/

File: C-Java/tp3/partieB.java
////////////////////////////////////////////
//  Travail pratique #3                   //
//  Auteur : Patrick Chartrand (20017273) //
//  Date : 23 avril 2020                  //
////////////////////////////////////////////

// PARTIE B
public class partieB {

    // fonction pour afficher le tableau
    static void afficher(char[] poste, double[] cafe, double[] age, int nbEmp)
    {
        String messageAv = "avant le triage";
        // message d'avant tri
        System.out.printf("Tableau 1 :\n%s\n", messageAv);
        System.out.printf("====================\n");
        System.out.printf("poste || cafe || age\n");
        System.out.printf("====================\n");
        // répartition des données
	    for(int i = 0; i < nbEmp; i++)
	    System.out.printf("  %c   ||  %.0f   ||  %.0f\n", poste[i], cafe[i], age[i]);
        System.out.printf("====================\n\n");
    }
    
    // fonction pour les calculs
    static int combien(char posteChoisi, char[] poste, int nbEmp)
    {
        int n = 0;
        for(int i = 0; i < nbEmp; i++)
        if (poste[i] == posteChoisi)
            n++;
        
        return n;
    }
    
    // fonction pour les moyennes
    static double moyenne(double[] tab, int nbElem)
    {
        double somme = 0.0;

        for(int i = 0; i < nbElem; i++)
        somme += tab[i];
   	    
   	    return somme / nbElem;
    }
    
    // fonction pour le triage
    static void trier(char poste[], double cafe[], double age[], int nbEmp)
    {
        int indMin, i, j;
        // variables temporaires
        double tempoAge, tempoCafe;
        char tempoPoste;
        // tri
        for (i = 0; i < nbEmp - 1; i++)
        {
            indMin = i;
            for (j = i + 1; j < nbEmp; j++)
            if (age[j] < age[indMin])
                indMin = j;
                if (indMin != i)
                {
                    tempoAge = age[i];
                    age[i] = age[indMin];
                    age[indMin] = tempoAge;
          
                    tempoCafe = cafe[i];
                    cafe[i] = cafe[indMin];
                    cafe[indMin] = tempoCafe;

                    tempoPoste = poste[i];
                    poste[i] = poste[indMin];
                    poste[indMin] = tempoPoste;
                } 
            }
            String messageAp = "après le triage (par ordre croissant d'âges)";
            // message d'après tri
            System.out.printf("Tableau 2 :\n%s\n", messageAp);
            System.out.printf("====================\n");
            System.out.printf("poste || cafe || age\n");
            System.out.printf("====================\n");
            // répartition des données
    	    for(i = 0; i < nbEmp; i++)
    	    System.out.printf("  %c   ||  %.0f   ||  %.0f\n", poste[i], cafe[i], age[i]);
            System.out.printf("====================\n\n");
        }

    // espace principal
    public static void main(String[] args)
    {
        // les données du tableau
        char poste[] = { 'P', 'P', 'O', 'A', 'P', 'A', 'P', 'P' };
        double [] cafe = {  3 ,  1 ,  6 ,  0 ,  5 ,  4 ,  0 ,  3  }, age = { 25 , 19 , 27 , 26 , 69 , 24 , 56 , 29  };

        // longueur du tableau
        int nbEmp = poste.length;
        
        // retour de fonctions
	    afficher(poste, cafe, age, nbEmp);
	    trier(poste, cafe, age, nbEmp);
	    
        // affichage de fonctions retournées ici
        System.out.printf("Nombre de programmeurs : %d\n", combien('P', poste, nbEmp));
        System.out.printf("Nombre d'opérateurs : %d\n", combien('O', poste, nbEmp));
        System.out.printf("Nombre d'analystes : %d\n", combien('A', poste, nbEmp));
        System.out.printf("Nombre de secrétaires : %d\n", combien('S', poste, nbEmp));
        System.out.printf("Moyenne d'âge parmi tous les employés : %.2f ans\n", moyenne(age, nbEmp));
        System.out.printf("Moyenne de consommation en café parmi tous les employés : %.2f\n", moyenne(cafe, nbEmp));
    }
    
}


/* EXÉCUTION

Tableau 1 :
avant le triage
====================
poste || cafe || age
====================
  P   ||  3   ||  25
  P   ||  1   ||  19
  O   ||  6   ||  27
  A   ||  0   ||  26
  P   ||  5   ||  69
  A   ||  4   ||  24
  P   ||  0   ||  56
  P   ||  3   ||  29
====================

Tableau 2 :
après le triage (par ordre croissant d'âges)
====================
poste || cafe || age
====================
  P   ||  1   ||  19
  A   ||  4   ||  24
  P   ||  3   ||  25
  A   ||  0   ||  26
  O   ||  6   ||  27
  P   ||  3   ||  29
  P   ||  0   ||  56
  P   ||  5   ||  69
====================

Nombre de programmeurs : 5
Nombre d'opérateurs : 1
Nombre d'analystes : 2
Nombre de secrétaires : 0
Moyenne d'âge parmi tous les employés : 34.38 ans
Moyenne de consommation en café parmi tous les employés : 2.75

*/